plugin-server/src/worker/ingestion/timestamps.ts

```diff
@@ -34,52 +34,72 @@ export function parseEventTimestamp(data: PluginEvent, callback?: IngestionWarni
         return DateTime.utc()
     }
 
+    return parsedTs
+}
+
+function handleTimestamp(
+    data: PluginEvent,
+    now: DateTime,
+    sentAt: DateTime | null,
+    callback?: IngestionWarningCallback
+): DateTime {
+    let parsedTs: DateTime = now
+    let timestamp: DateTime = now
+
+    if (data['timestamp']) {
+        timestamp = parseDate(data['timestamp'])
+
+        if (!sentAt || !timestamp.isValid) {
+            return timestamp
+        }
+
+        // To handle clock skew between the client and server, we attempt
+        // to compute the skew based on the difference between the
+        // client-generated `sent_at` and the server-generated `now`
+        // filled by the capture endpoint.
+        //
+        // We calculate the skew as:
+        //
+        //      skew = sent_at - now
+        //
+        // And adjust the timestamp accordingly.
+
+        // sent_at - timestamp == now - x
+        // x = now + (timestamp - sent_at)
+        try {
+            // timestamp and sent_at must both be in the same format: either both with or both without timezones
+            // otherwise we can't get a diff to add to now
+            parsedTs = now.plus(timestamp.diff(sentAt))
+        } catch (error) {
+            status.error('⚠️', 'Error when handling timestamp:', { error: error.message })
+            Sentry.captureException(error, { extra: { data, now, sentAt } })
+            return timestamp
+        }
+    }
+
+    if (data['offset']) {
+        parsedTs = now.minus(Duration.fromMillis(data['offset']))
+    }
+
+    const nowDiff = parsedTs.toUTC().diff(now).toMillis()
+
     // Events in the future would indicate an instrumentation bug, lets' ingest them
     // but publish an integration warning to help diagnose such issues.
-    if (now.isValid && parsedTs.toUTC().diff(now).toMillis() > FutureEventHoursCutoffMillis) {
+    if (nowDiff > FutureEventHoursCutoffMillis) {
         callback?.('event_timestamp_in_future', {
             timestamp: data['timestamp'] ?? '',
             sentAt: data['sent_at'] ?? '',
             offset: data['offset'] ?? '',
             now: data['now'],
             result: parsedTs.toISO(),
+            eventUuid: data['uuid'],
+            eventName: data['event'],
```

Comment:
Extend https://github.com/PostHog/posthog/blob/77fd7ffad6b24de5fc8eab2c355d860d0cd8d06b/frontend/src/scenes/data-management/ingestion-warnings/IngestionWarningsView.tsx#L87 (could be a separate follow-up PR), specifically linking to an event search for this uuid would be super helpful to understand the error/problem better for users.